src/librustc/traits/error_reporting/suggestions.rs

```diff
@@ -0,0 +1,1695 @@
+                ty::Dynamic(predicates, _) => {
+                    // The `dyn Trait` is not object safe, do not suggest `Box<dyn Trait>`.
+                    is_object_safe = predicates.principal_def_id().map_or(true, |def_id| {
+                        !object_safety_violations(self.tcx, def_id).is_empty()
+                    })
+                }
+                // We only want to suggest `impl Trait` to `dyn Trait`s.
+                // For example, `fn foo() -> str` needs to be filtered out.
+                _ => return false,
+            }
+
+            let ret_ty = if let hir::FunctionRetTy::Return(ret_ty) = sig.decl.output {
+                ret_ty
+            } else {
+                return false;
+            };
+
+            // Use `TypeVisitor` instead of the output type directly to find the span of `ty` for
+            // cases like `fn foo() -> (dyn Trait, i32) {}`.
+            // Recursively look for `TraitObject` types and if there's only one, use that span to
+            // suggest `impl Trait`.
+
+            // Visit to make sure there's a single `return` type to suggest `impl Trait`,
+            // otherwise suggest using `Box<dyn Trait>` or an enum.
+            let mut visitor = ReturnsVisitor(vec![]);
+            visitor.visit_body(&body);
+
+            let tables = self.in_progress_tables.map(|t| t.borrow()).unwrap();
+
+            let mut all_returns_conform_to_trait = true;
+            let mut all_returns_have_same_type = true;
+            let mut last_ty = None;
+            if let Some(ty_ret_ty) = tables.node_type_opt(ret_ty.hir_id) {
+                let cause = ObligationCause::misc(ret_ty.span, ret_ty.hir_id);
+                let param_env = ty::ParamEnv::empty();
+                if let ty::Dynamic(predicates, _) = &ty_ret_ty.kind {
+                    for expr in &visitor.0 {
+                        if let Some(returned_ty) = tables.node_type_opt(expr.hir_id) {
+                            all_returns_have_same_type &=
+                                Some(returned_ty) == last_ty || last_ty.is_none();
+                            last_ty = Some(returned_ty);
+                            for predicate in predicates.iter() {
+                                let pred = predicate.with_self_ty(self.tcx, returned_ty);
+                                let obl = Obligation::new(cause.clone(), param_env, pred);
+                                all_returns_conform_to_trait &= self.predicate_may_hold(&obl);
+                            }
+                        }
+                    }
+                }
+            } else {
+                // We still want to verify whether all the return types conform to each other.
+                for expr in &visitor.0 {
+                    let returned_ty = tables.node_type_opt(expr.hir_id);
+                    all_returns_have_same_type &= last_ty == returned_ty || last_ty.is_none();
+                    last_ty = returned_ty;
+                }
```

Comment:
This logic still exists in both branches and could be extracted out. Moreover, it would be more self contained to reformulate this as a fold:
```rust
let (last_ty, all_returns_have_same_type) = visitor.0.iter()
    .filter_map(|expr| tables.node_type_opt(expr.hir_id))
    .fold((None, true), |(last_ty, mut same), returned_ty| {
        same &= last_ty.map_or(true, |ty| ty == returned_ty);
        (Some(returned_ty|), same)
    });
```

Reply:
I don't see how to deduplicate the two branches without making the first one duplicate not only a bit of logic but also actions to be able to check the predicates.

Reply:
The fold is less important than the duplication, but here's how I'd write it:

```rust
        let ret_types = visitor
            .0
            .iter()
            .filter_map(|expr| tables.node_type_opt(expr.hir_id));

        let (last_ty, all_returns_have_same_type) =
            ret_types
                .clone()
                .fold((None, true), |(last_ty, mut same), returned_ty| {
                    same &= last_ty.map_or(true, |ty| ty == returned_ty);
                    (Some(returned_ty), same)
                });

        let all_returns_conform_to_trait = match ret_ty {
            Some(TyS {
                kind: ty::Dynamic(predicates, _),
            }) => {
                let cause = ObligationCause::misc(ret_ty.span, ret_ty.hir_id);
                let param_env = ty::ParamEnv::empty();
                ret_types
                    .cartesian_product(predicates.iter())
                    .all(|(returned_ty, predicate)| {
                        let pred = predicate.with_self_ty(self.tcx, returned_ty);
                        let obl = Obligation::new(cause.clone(), param_env, pred);
                        self.predicate_may_hold(&obl)
                    })
            }
            _ => true,
        };
```

The way ^--- is written, it's two independent algorithms whereas you've entangled them.

Anyways; If you don't want to disentangle them, r=me with the bare trait wording fixed.

Reply:
Adapted slightly, but done.